### PR TITLE
Large training batches on limited GPU hardware

### PR DESCRIPTION
**Pull request by fstahlberg**
This PR adds a LargebatchAdam optimizer, which accumulates gradients over n batches and applies the Adam learning rule every n batches on the accumulated gradients. This makes it possible to arbitrarily increase the effective batch size / number of GPUs at cost of more training iterations. This technique is useful if the number of physical GPUs is limited or the GPU memory does not allow to increase the batch size any further. Large batch / multi-GPU training is often important for Transformer training as reported [here](https://github.com/tensorflow/tensor2tensor/issues/444#issuecomment-384102731).

### CODE REVIEW

**Comment by stefan-it** [comment]
@fstahlberg Looks really cool! Could you provide some results of experiments, e.g. translation problem with BLEU score and a comparison with one vs. multiple GPUs?

**Comment by fstahlberg** [comment]
Closing this in favour of #754 as some commits were submitted with an invalid e-mail address which caused problems with the Google CLA, and my attempts to amend the e-mail address in previous commits were not successful.
